lib/handleRequest.js

```diff
@@ -88,8 +88,10 @@ function preHandlerCallback (err, reply) {
       .then((payload) => {
         // this is for async functions that
         // are using reply.send directly
-        if (payload !== undefined || (reply.res.statusCode === 204 && !reply.sent)) {
+        if (payload !== undefined || reply.res.statusCode === 204) {
           reply.send(payload)
+        } else if (!reply.sent) {
+          reply.res.log.error(new Error(`Promise may not be fulfilled with 'undefined' when statusCode is not 204`))
```

Comment:
Can you update the if guard with a more strict check? `reply.send === false` would be perfect.

I would log this as `warn`, as we are doing here:
https://github.com/fastify/fastify/blob/5fefda9dd2f882e807b7ae736d39be4860c13d6b/lib/reply.js#L36-L39

Reply:
I also wanted to log it as warn but if you take a closer look in this case the response will hang forever.

Reply:
I know, but the user will see this log (hopefully) only during development, so I think it is better to use a `warn`, as we did in `reply.js` or [here](https://github.com/fastify/fastify/blob/master/fastify.js#L719-L739).
It's for consistency :)

Reply:
Would agree but this isn't false / positive case and indicate a serious issue in your code, thus it's really an error.

Reply:
This case only happens when the user resolve the promise with `undefined` with statusCode != 204. This indicate a wrong use.